bundles/config/org.eclipse.smarthome.config.discovery.usbserial/src/main/java/org/eclipse/smarthome/config/discovery/usbserial/internal/UsbSerialDiscoveryService.java

```diff
@@ -0,0 +1,159 @@
+ * USB devices but rather use a serial port for the communication, where the serial port is provided by an operating
+ * system driver outside the scope of Eclipse SmartHome. Examples for such USB devices are USB dongles that provide
+ * access to wireless networks, like, e.g., Zigbeee or Zwave dongles.
+ * <p/>
+ * This discovery service provides functionality for discovering added and removed USB devices and the corresponding
+ * serial ports. The actual {@link DiscoveryResult}s are then provided by {@link UsbSerialDiscoveryParticipant}s, which
+ * are called by this discovery service whenever new devices are detected or devices are removed. Such
+ * {@link UsbSerialDiscoveryParticipant}s should be provided by bindings accessing USB devices via a serial port.
+ * <p/>
+ * This discovery service requires a component implementing the interface {@link UsbSerialDiscovery}, which performs the
+ * actual serial port and USB device discovery (as this discovery might differ depending on the operating system).
+ *
+ * @author Henning Sudbrock - initial contribution
+ */
+@NonNullByDefault
+@Component(immediate = true, service = { DiscoveryService.class,
+        UsbSerialDiscoveryService.class }, configurationPid = "discovery.usbserial")
+public class UsbSerialDiscoveryService extends AbstractDiscoveryService implements UsbSerialDiscoveryListener {
+
+    private final Logger logger = LoggerFactory.getLogger(UsbSerialDiscoveryService.class);
+
+    private final Set<UsbSerialDiscoveryParticipant> discoveryParticipants = new CopyOnWriteArraySet<>();
+
+    @Nullable
+    private volatile UsbSerialDiscovery usbSerialDiscovery;
+
+    public UsbSerialDiscoveryService() {
+        super(5);
+    }
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    protected void addUsbSerialDiscoveryParticipant(UsbSerialDiscoveryParticipant participant) {
+        this.discoveryParticipants.add(participant);
+    }
+
+    protected void removeUsbSerialDiscoveryParticipant(UsbSerialDiscoveryParticipant participant) {
+        this.discoveryParticipants.remove(participant);
+    }
+
+    @Reference(cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.DYNAMIC)
+    protected synchronized void setUsbSerialDiscovery(UsbSerialDiscovery usbSerialDiscovery) {
+        UsbSerialDiscovery currentUsbSerialDiscovery = this.usbSerialDiscovery;
+        if (currentUsbSerialDiscovery != null) {
+            unsetUsbSerialDiscovery(currentUsbSerialDiscovery);
+        }
+
+        this.usbSerialDiscovery = usbSerialDiscovery;
+        usbSerialDiscovery.registerDiscoveryListener(this);
+        if (isBackgroundDiscoveryEnabled()) {
+            usbSerialDiscovery.startBackgroundScanning();
+        }
+    }
+
+    protected synchronized void unsetUsbSerialDiscovery(UsbSerialDiscovery usbSerialDiscovery) {
+        if (this.usbSerialDiscovery == usbSerialDiscovery) {
```

Comment:
Does it make sense to have this method parameter and do this check here? Usually from an `unset...` method I do expect that it really unsets my variable. In your case the variable is only unset if it has the same value as the method parameter...

Reply:
I introduced this due to @amitjoy 's review comment above (https://github.com/eclipse/smarthome/pull/5315#discussion_r177174276). The reason is that the `usbSerialDiscovery` is an optional dynamic dependency, and as indicated by @amitjoy , the OSGi container does not make any guarantees on the call order of the set and unset methods when there are more than one `usbSerialDiscovery` services available. I imagine that this could happen when changing to another implementation of `UsbSerialDiscovery` during bundle update, e.g., when replacing the current `usbSerialDiscovery` with an improved one. Then the OSGi container could first set the new service, and only unset the old one afterwards. Hence, as suggested by @amitjoy, the unset method does not unconditionally unset the current `usbSerialDiscovery`, but only unsets the `usbSerialDiscovery` passed as method parameter.

Reply:
I just discussed briefly with @sjka - it is probably possible to simple use a static dependency here - I will check if I can improve the code this way.